File: whatifi.py
#!/usr/local/bin/python3

class income():

    def __init__(self):
        self.value = 0
        self.date_start = 0
        self.date_end = -1

    def value_for_date(self,date):
        if date >= self.date_start:
            if date <= self.date_end or self.date_end == -1:
                return self.value

        return 0

    def total_value(self,start,end):
        tv = 0
        for i in range(start,end):
            tv += self.value_for_date(i)

        return tv

class person():

    def __init__(self):
        self.values = []

    def total_value(self,start,end):
        tv = 0
        for value in self.values:
            tv += value.total_value(start,end)

        return tv

class scenario():

    def __init__(self):
        self.scenarios = []
        self.is_selected = False
        self.total_value = 0


    def


def main():
    you = person()

    yi1 = income()
    yi1.value = 10
    yi1.date_start = 2

    you.values.append(yi1)

    tax_return = income()
    tax_return.value = 100
    tax_return.date_end = 0
    you.values.append(tax_return)

    partner = person()

    pi1 = income()
    pi1.value = 20
    pi1.date_start = 1

    partner.values.append(yi1)
    print("partner: %i" % partner.total_value(0,10))
    you.values.append(partner)

    ch1 = person()
    ci1 = income()
    ci1.value = -5
    ci1.date_start = 0
    ch1.values.append(ci1)

    ch2 = person()
    ci2 = income()
    ci2.value = -10
    ci2.date_start = 0
    ch2.values.append(ci2)

    you.values.append(ch1)
    you.values.append(ch2)

    print("you: %i" % you.total_value(0,10))

if __name__ == "__main__":
    main()

File: WhatIfI.js
class Income {

    constructor() {
        this._value = 0;
        this._dateStart = 0;
        this._dateEnd = -1;
    }

    get value() {
        return this._value;
    }

    set value(newValue) {
        this._value = newValue;
    }

    set dateStart(newDateStart) {
        this._dateStart = newDateStart;
    }

    set dateEnd(newDateEnd) {
        this._dateEnd = newDateEnd;
    }

    valueForDate(date) {
        if (date >= this._dateStart) {
            if (date <= this._dateEnd || this._dateEnd === -1) {
                return this._value;
            }
        }
        return 0;
    }

    totalValue(start, end) {
        let tv = 0;
        for(let i = start; i <= end; i++) {
            tv += this.valueForDate(i);
        }
        return tv;
    }

}

class Person {
    constructor() {
        this._values = [];
    }

    get values() {
        return this._values;
    }

    totalValue(start, end) {
        let tv = 0;
        this._values.forEach(function(value) {
            tv += value.totalValue(start, end);
        });
        return tv;
    }

}

const main = () => {
    const you = new Person();
    var yourIncome1 = new Income();

    yourIncome1.value = 10;
    yourIncome1.dateStart = 2;

    you.values.push(yourIncome1);

    console.log(`Your income 1: ${you.totalValue(0, 10)}`);

    const taxReturn = new Income();
    taxReturn.value = 100;
    taxReturn.dateEnd = 0;
    you.values.push(taxReturn);

    const partner = new Person();

    const partnerIncome1 = new Income();
    partnerIncome1.value = 20;
    partnerIncome1.dateStart = 1;

    partner.values.push(partnerIncome1);
    console.log(`Partner: ${partner.totalValue(0, 10)}`);
    you.values.push(partner);

    const child1 = new Person();

    const childIncome1 = new Income();
    childIncome1.value = -5;
    childIncome1.dateStart = 0;
    child1.values.push(childIncome1);

    console.log(`Child income 1: ${child1.totalValue(0, 10)}`);

    const child2 = new Person();

    const childIncome2 = new Income();
    childIncome2.value = -10;
    childIncome2.dateStart = 0;
    child2.values.push(childIncome2);

    console.log(`Child income 2: ${child2.totalValue(0, 10)}`);

    you.values.push(childIncome1);
    you.values.push(childIncome2);

    console.log(`You: ${you.totalValue(0, 10)}`);

}

main();